File: 2022/day2/day2.rs
//! [Day 2: Rock Paper Scissors](https://adventofcode.com/2022/day/2)

const VALUE_ROCK: u32 = 1;
const VALUE_PAPER: u32 = 2;
const VALUE_SCISSORS: u32 = 3;

const SHOULD_LOSE: u32 = 1;
const SHOULD_DRAW: u32 = 2;

const ROUND_OUTCOME_DRAW: u32 = 3;
const ROUND_OUTCOME_WIN: u32 = 6;

struct Puzzle {
    guide: Vec<(u32, u32)>,
}

impl Puzzle {
    const fn new() -> Self {
        Self { guide: Vec::new() }
    }

    fn configure(&mut self, data: &str) {
        let lines = data.trim().lines().collect::<Vec<_>>();

        for strategy in lines {
            let shapes = strategy.split(' ').collect::<Vec<_>>();
            let opponent = match shapes[0].parse::<char>().unwrap() {
                'A' => VALUE_ROCK,
                'B' => VALUE_PAPER,
                'C' => VALUE_SCISSORS,
                _ => panic!("bad input"),
            };
            let you = match shapes[1].parse::<char>().unwrap() {
                'X' => 1,
                'Y' => 2,
                'Z' => 3,
                _ => panic!("bad input"),
            };
            self.guide.push((opponent, you));
        }
    }

    fn part1(&self) -> u32 {
        let mut result = 0;
        for strategy in &self.guide {
            let (opponent, you) = *strategy;

            // Always win what has been played
            result += you;

            // Take advantage of shapes sorting to check win conditions
            if opponent + 1 == you || you + 2 == opponent {
                result += ROUND_OUTCOME_WIN;
            } else if opponent == you {
                result += ROUND_OUTCOME_DRAW;
            }
        }
        result
    }

    fn part2(&self) -> u32 {
        let mut result = 0;
        for strategy in &self.guide {
            let (opponent, you) = *strategy;

            if you == SHOULD_LOSE {
                if opponent == VALUE_ROCK {
                    result += VALUE_SCISSORS;
                } else {
                    // Take advantage of shapes sorting to know what to play
                    result += opponent - 1;
                }
            } else if you == SHOULD_DRAW {
                result += ROUND_OUTCOME_DRAW + opponent;
            } else {
                result += ROUND_OUTCOME_WIN;
                if opponent == VALUE_SCISSORS {
                    result += VALUE_ROCK;
                } else {
                    // Take advantage of shapes sorting to know what to play
                    result += opponent + 1;
                }
            }
        }
        result
    }
}

/// Solve the puzzle with the user input
fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[test]
fn test01() {
    let mut puzzle = Puzzle::new();
    puzzle.configure(&aoc::load_input_data("test.txt"));
    assert_eq!(puzzle.part1(), 15);
    assert_eq!(puzzle.part2(), 12);
}

File: 2022/day2/day2.py
#!/usr/bin/env python3

import sys

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = open(filename).read()

SCORE_WIN = 6
SCORE_DRAW = 3
SCORE_LOOSE = 0

ROCK = 1
PAPER = 2
SCISSORS = 3

NEED_TO_LOOSE = 1
NEED_TO_DRAW = 2
NEED_TO_WIN = 3

part1 = 0
part2 = 0

for i in data.splitlines():
    opponent, me = i.split()

    opponent = ord(opponent) - ord("A") + 1
    me = ord(me) - ord("X") + 1

    # part 1
    if opponent == me:
        part1 += me + SCORE_DRAW

    elif opponent == ROCK and me == SCISSORS:  # the rock breaks the scissors
        part1 += me + SCORE_LOOSE
    elif opponent == SCISSORS and me == PAPER:  # the scissors cut the paper
        part1 += me + SCORE_LOOSE
    elif opponent == PAPER and me == ROCK:  # the paper covers the scissors
        part1 += me + SCORE_LOOSE

    else:
        part1 += me + SCORE_WIN

    # part 2
    if me == NEED_TO_DRAW:
        part2 += opponent + SCORE_DRAW

    elif me == NEED_TO_LOOSE:
        if opponent == ROCK:
            part2 += SCISSORS + SCORE_LOOSE
        elif opponent == PAPER:
            part2 += ROCK + SCORE_LOOSE
        elif opponent == SCISSORS:
            part2 += PAPER + SCORE_LOOSE
    else:
        if opponent == ROCK:
            part2 += PAPER + SCORE_WIN
        elif opponent == PAPER:
            part2 += SCISSORS + SCORE_WIN
        elif opponent == SCISSORS:
            part2 += ROCK + SCORE_WIN

print(part1)
print(part2)
